### PR TITLE
UC-1246 Add example of generating a report

### PR DESCRIPTION
**Pull request by sedwards2009**
UC-1246

Note: I had to update scopes on production too.

### CODE REVIEW

File: typescript/src/digital-factory.ts [hunk 1]
```diff
@@ -303,4 +303,88 @@ export class DigitalFactoryDemo {
...
+        let status: ReportStatus = response.data;
+        const reportID = status.report_id;
+        while (status.status === 'pending' || status.status === 'processing') {
+            print('Waiting one second');
+            await new Promise((resolve) => setTimeout(resolve, 1000));
+
+            print('Checking the status of the report');
+            const statusResponse = <{ data: ReportStatus }> await this.httpGetDigitalFactory(`${API_ROOT_URL}/report/v1/reports/${reportID}`);
+            status = statusResponse.data;
+        }
```
Comment (DanielSchiavini) [inline typescript/src/digital-factory.ts:333]: Await in a loop is usually bad practice... And the code is repeating a lot. I feel like the polling here should be extracted away from the main logic. e.g.
```suggestion
        const result = await poll(
            () => this.httpGetDigitalFactory(`${API_ROOT_URL}/report/v1/reports/${reportID}`),
            response => !['pending', 'processing'].includes(response.status)
        )
        const status: ReportStatus = result.status;
        const reportID = status.report_id;
```

File: typescript/src/digital-factory.ts [hunk 2]
```diff
@@ -303,4 +303,88 @@ export class DigitalFactoryDemo {
...
+        if (status.status !== 'success') {
+            print('Something went wrong while generating the report.', status);
+            return null;
```
Comment (DanielSchiavini) [inline typescript/src/digital-factory.ts:336]: Shouldn't this be an exception?

File: typescript/src/digital-factory.ts [hunk 3]
```diff
@@ -303,4 +303,88 @@ export class DigitalFactoryDemo {
...
+        }
+        if (status.status !== 'success') {
+            print('Something went wrong while fetching the webcam image.', status);
+            return null;
+        }
+        return status.status_details.value.image_url;
+    }
+}
+
+interface ReportStatus {
```
Comment (DanielSchiavini) [inline typescript/src/digital-factory.ts:367]: why aren't we using stardust-api-clients?

File: typescript/src/main.ts [hunk 4]
```diff
@@ -46,6 +46,23 @@ async function main(): Promise<void> {
     } else {
         print('No projects found.\n');
     }
+
+    print('');
```
Comment (DanielSchiavini) [inline typescript/src/main.ts:50]:
```suggestion
```

File: typescript/src/main.ts [hunk 5]
```diff
@@ -46,6 +46,23 @@ async function main(): Promise<void> {
...
+    const clusters = await demo.getClusters();
+    if (clusters.length !== 0) {
```
Comment (DanielSchiavini) [inline typescript/src/main.ts:52]: this if is not necessary and may be safely removed

Comment (DanielSchiavini) [inline typescript/src/main.ts:52]: this if is not necessary and may be safely removed

Comment (DanielSchiavini) [inline typescript/src/main.ts:52]: this if is not necessary

File: typescript/src/main.ts [hunk 6]
```diff
@@ -46,6 +46,23 @@ async function main(): Promise<void> {
...
+        const onlineClusters = clusters.filter((c) => c.is_online);
+        if (onlineClusters.length !== 0) {
+            const webcamClusterId = onlineClusters[0].cluster_id;
+            const webcamPrinterId = onlineClusters[0].host_printer.uuid;
+            const imageUrl = await demo.getWebcamImage(webcamClusterId, webcamPrinterId);
+            print(`Received webcam image URL: ${imageUrl}`);
+        }
+    }
+
+    print('');
```
Comment (DanielSchiavini) [inline typescript/src/main.ts:62]:
```suggestion
```

File: typescript/src/digital-factory.ts (hunk 1 above)
Comment (sedwards2009) [reply typescript/src/digital-factory.ts:333]: There is nothing wrong with an `await` in a loop. I don't know where that idea is coming from. You could say that this loop should give up after a certain amount of time though. But the purpose of the code is to be readable first and to teach what the happy flow looks like. Introducing a `poll()` function makes sense and leans DRY, but gets in the way of teaching. So I wouldn't do it.

File: typescript/src/digital-factory.ts (hunk 2 above)
Comment (sedwards2009) [reply typescript/src/digital-factory.ts:336]: Possibly, but it is demo code.

File: typescript/src/digital-factory.ts (hunk 3 above)
Comment (sedwards2009) [reply typescript/src/digital-factory.ts:367]: Our customers don't have access to the `startdust-api-clients` package.

File: typescript/src/main.ts (hunk 6 above)
Comment (sedwards2009) [reply typescript/src/main.ts:62]: The output looks crappy without and extra line there.

**Review by DanielSchiavini — APPROVED** [review]
IMO we should really point the users to a generated API client

**Pull request merged**